hazelcast/src/test/java/com/hazelcast/spi/impl/sequence/FailFastCallIdSequenceTest.java

```diff
@@ -0,0 +1,148 @@
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.RequireAssertEnabled;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.CountDownLatch;
+
+import static com.hazelcast.spi.OperationAccessor.setCallId;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class FailFastCallIdSequenceTest extends HazelcastTestSupport {
+
+    FailFastCallIdSequence sequence = new FailFastCallIdSequence(100);
+
+    @Test
+    public void test() {
+        assertEquals(0, sequence.getLastCallId());
+        assertEquals(100, sequence.getMaxConcurrentInvocations());
+    }
+
+    @Test
+    public void whenNext_thenSequenceIncrements() {
+        // regular operation
+        testNext(new DummyOperation());
+
+        // backup-aware operation
+        testNext(new DummyBackupAwareOperation());
+
+        // urgent operation
+        testNext(new DummyPriorityOperation());
+    }
+
+    private void testNext(Operation operation) {
+        long oldSequence = sequence.getLastCallId();
+        long result = nextCallId(sequence, operation.isUrgent());
+        assertEquals(oldSequence + 1, result);
+        assertEquals(oldSequence + 1, sequence.getLastCallId());
+    }
+
+    @Test (expected = HazelcastOverloadException.class)
+    public void next_whenNoCapacity_thenThrowException() throws InterruptedException {
+        sequence = new FailFastCallIdSequence(1);
+        final long oldLastCallId = sequence.getLastCallId();
+
+        final CountDownLatch nextCalledLatch = new CountDownLatch(1);
+        spawn(new Runnable() {
+            @Override
+            public void run() {
+                DummyBackupAwareOperation op = new DummyBackupAwareOperation();
+                long callId = nextCallId(sequence, op.isUrgent());
+                setCallId(op, callId);
+                nextCalledLatch.countDown();
+                sleepSeconds(3);
+                sequence.complete();
+            }
+        });
+
+        nextCalledLatch.await();
+
+        nextCallId(sequence, false);
+    }
+
+    @Test
+    public void next_whenNoCapacity_thenBlockTillTimeout() {
+        sequence = new FailFastCallIdSequence(1);
+
+        // first invocation consumes the available call ID
+        nextCallId(sequence, false);
+
+        long oldLastCallId = sequence.getLastCallId();
+        try {
+            sequence.next();
+            fail();
+        } catch (HazelcastOverloadException e) {
+            // expected
+        }
+
+        assertEquals(oldLastCallId, sequence.getLastCallId());
+    }
+
+    @Test
+    public void when_overCapacityButPriorityItem_then_noBackpressure() {
```

Comment:
There will never by any back pressure. So the name of this method is confusing.

whenAllSlotsTaken_andForceNext_thenExtraCallIdGranted

Something like this.

Reply:
OK, changed.